File: packages/render/src/crawl.ts
import { existsSync, readFileSync } from "node:fs";
import { resolve } from "node:path";
import { isMarkdownPath, localMarkdownLink } from "@portfolio/core";
import { markdownToAst } from "./pandoc.ts";
import { astLinks } from "./links.ts";

export interface CrawlResult {
  roots: string[];
  /** Every file to import, sorted. Includes the roots. */
  files: string[];
  /** Absolute path -> the local Markdown files it links to. */
  edges: Map<string, string[]>;
}

/**
 * Starting from root files, follows links to local Markdown files (when
 * `recursive`) and returns the closure. Throws when a linked file is missing,
 * naming the file that linked it. HTML roots are included but never crawled.
 */
export function crawlMarkdown(rootPaths: string[], options: { recursive?: boolean; bin?: string; read?: (path: string) => string } = {}): CrawlResult {
  const read = options.read ?? (path => readFileSync(path, "utf8"));
  const roots = rootPaths.map(path => resolve(path));
  const seen = new Set<string>();
  const edges = new Map<string, string[]>();
  const queue = [...roots].sort();
  while (queue.length) {
    const path = queue.shift()!;
    if (seen.has(path)) continue;
    if (!existsSync(path)) throw new Error(`linked Markdown not found: ${path}`);
    seen.add(path);
    if (!options.recursive || !isMarkdownPath(path)) continue;
    const children: string[] = [];
    for (const target of astLinks(markdownToAst(read(path), options.bin))) {
      const child = localMarkdownLink(path, target);
      if (!child) continue;
      if (!existsSync(child.sourcePath)) throw new Error(`linked Markdown not found: ${child.sourcePath} (from ${path})`);
      children.push(child.sourcePath);
      if (!seen.has(child.sourcePath)) queue.push(child.sourcePath);
    }
    edges.set(path, [...new Set(children)]);
    queue.sort();
  }
  return { roots, files: [...seen].sort(), edges };
}

File: lib/pf/crawl.rb
# frozen_string_literal: true

require_relative 'detect'
require_relative 'render'

module Pf
  module Crawl
    module_function

    def crawl_markdown(root_paths, recursive: false, bin: nil, read: nil)
      reader = read || ->(path) { File.read(path, encoding: 'UTF-8') }
      roots = root_paths.map { |path| File.expand_path(path) }
      seen = {}
      edges = {}
      queue = roots.sort

      until queue.empty?
        path = queue.shift
        next if seen[path]
        raise Pf::Error, "linked Markdown not found: #{path}" unless File.exist?(path)

        seen[path] = true
        next unless recursive && Pf::Detect.markdown_path?(path)

        children = []
        Pf::Render.ast_links(Pf::Render.markdown_to_ast(reader.call(path), bin: bin)).each do |target|
          child = Pf::Detect.local_markdown_link(path, target)
          next unless child
          child_path = child['sourcePath']
          raise Pf::Error, "linked Markdown not found: #{child_path} (from #{path})" unless File.exist?(child_path)

          children << child_path
          queue << child_path unless seen[child_path]
        end
        edges[path] = children.uniq
        queue.sort!
      end

      { 'roots' => roots, 'files' => seen.keys.sort, 'edges' => edges }
    end
  end
end
